crates/aranya-daemon/src/policy.md

```diff
@@ -832,51 +1104,61 @@ command RevokeOwner{
         check team_exists()
 
         let author = get_valid_device(envelope::author_id(envelope))
-        let device = get_valid_device(this.device_id)
+        let target = get_valid_device(this.device_id)
 
-        // Owner can only revoke the role from itself.
-        check author.device_id == this.device_id
-        // Check that the device is an Owner.
-        check is_owner(author.role)
+        check device_can_execute_op(author.device_id, "RevokeRole")
+        check author_dominates_target(author.device_id, target.device_id)
+
+        // Query role.
+        let role = check_unwrap query Role[role_id: this.role_id]
 
         finish {
-            revoke_role(device)
+            delete AssignedRole[role_id: role.role_id, device_id: target.device_id]
 
-            emit OwnerRevoked {
-                device_id: this.device_id,
+            // Return revoked role info.
+            emit RoleRevoked {
+                device_id: target.device_id,
+                role_id: role.role_id,
+                name: role.name,
+                author_id: author.device_id,
             }
         }
     }
 }
 
-// Revokes the specified role from the device. This automatically sets their role to Member instead.
-finish function revoke_role(device struct Device) {
-    update Device[device_id: device.device_id]=>{
-        role: device.role,
-        sign_key_id: device.sign_key_id,
-        enc_key_id: device.enc_key_id,
-        } to {
-            role: Role::Member,
-            sign_key_id: device.sign_key_id,
-            enc_key_id: device.enc_key_id,
-            }
+// A role was revoked from a device on the team.
+effect RoleRevoked {
+    // ID of device the role was revoked from.
+    device_id id,
+    // ID of the role.
+    role_id id,
+    // Name of the role.
+    name string,
+    // ID of the device that revoked the role.
+    author_id id,
 }
 ```
 
-### RevokeAdmin
+## AssignRoleOp
 
-Revoke the `Admin` role from a device.
+Assign permission to execute a certain operation to the role.
+Note: currently each operation is only allowed to be executed by one role.
 
 ```policy
-// The Admin role was revoke from A device.
-effect AdminRevoked {
-    device_id id,
+
+action assign_operation_to_role(role_id id, op string) {
```

Comment:
It's the other way around. In order to perform an operation you must have been assigned a role. This should really be `require_role_for_operation`.

The language needs to be cleaned up elsewhere in the policy file.